### PR TITLE
Lua: add option to put backpressure when script executes httpCall

### LINKED ISSUE #10076: Lua filter buffering while making async HttpCall
Hi,

I am using Envoy in the following scenario :

Single listener with a single lua filter.
The lua filter makes a HTTP call to a third party API, extracts some data from the response and modify upstream headers accordingly. This is what it looks like :

```yaml
static_resources:
  listeners:
  - per_connection_buffer_limit_bytes: 1024
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 8080
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
          stream_idle_timeout: 0s
          codec_type: auto
          stat_prefix: storage-api
          server_name: "storage"
          route_config:
            name: s3_route
            virtual_hosts:
            - name: upload
              domains:
              - "*"
              routes:
              - match:
                  regex: "^.*$"
                  headers:
                   - name: ":method"
                     exact_match: "PUT"
                route:
                  cluster: s3
                  timeout: 0s
          http_filters:
          - name: envoy.lua
            typed_config:
              "@type": type.googleapis.com/envoy.config.filter.http.lua.v2.Lua
              inline_code: |
                function envoy_on_request(request_handle)

                  request_handle:logDebug("- - - - ON REQUEST UPLOAD - - - -")
                  
                  -- third party API call
                  local headers, body = request_handle:httpCall(
                  "thirdparty",
                  {
                   [":method"] = "GET",
                   [":path"] = "/v2/5e480bb3300000e539294a5e?mocky-delay=10s",
                   [":authority"] = "www.mocky.io"
                  },
                  nil,
                  65000)
                  request_handle:logDebug("mocky responded with ")
                  request_handle:logDebug(headers[":status"])
                  request_handle:logDebug(body)
                  

                  -- change upstream header with some information from the async http call
                  request_handle:headers():add("someHeader", body)
                  
                  
                  request_handle:logDebug("- - - - ON REQUEST UPLOAD END - - - -")

                end
          - name: envoy.router
            typed_config: {}
  clusters:
  - name: thirdparty
    connect_timeout: 10.0s
    type: logical_dns 
    dns_lookup_family: V4_ONLY
    lb_policy: round_robin
    load_assignment:
      cluster_name: thirdparty
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: www.mocky.io
                port_value: 80
  - name: s3
    connect_timeout: 10.0s
    type: strict_dns
    hosts:
      socket_address:
        address: 10.0.2.15
        port_value: 9000
  admin:
    access_log_path: "/dev/null"
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 8001

```

Since this API will be used to upload large files, I set the `per_connection_buffer_limit_bytes: 1024` because I don't want Envoy to run OOM.

The problem is that I get a 413 error if the buffer is full before the HTTP call has terminated :

```
[2020-02-16 13:21:24.758][6690][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1932] [C0][S7481487801671065876] request data too large watermark exceeded
[2020-02-16 13:21:24.758][6690][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1167] [C0][S7481487801671065876] Sending local reply with details request_payload_too_large
```

If I increase the buffer limit it works fine, but then the whole file is buffered by Envoy, which I want to avoid.

Is there a way to wait for the HTTP call to finish without buffering anything ? Putting the client on hold and then resuming streaming mode ? Watermarking instead of buffering would be perfect in this case.

Or, is there any alternative without LUA filter to what I'm trying to achieve here ?

EDIT : Subsidiary question about memory management :
When a request, or part of it, is buffered by Envoy, how and when is it cleared from memory ?
From what I've seen, when a request is buffered, it eventually raises the allocated memory size of the process. The process memory remains allocated but if it needs to buffer another request of the same size, it will reuse that allocated memory and won't use more.
i.e. the memory print of the process is equal to the maximum amount of data envoy had to buffer at one point.
Am I right ? If so, is the memory released at some point ? Like with a garbage-collector ?

### PR DESCRIPTION
**Pull request by cpakulski**
Commit Message:
add option to put backpressure when Lua script executes httpCall
Additional Description:
PR addresses a situation when upon receiving request headers, Lua script issues httpCall and backpressure should be put on the downstream client to slow down sending body, which size may exceed maximum allowed buffer size per connection.
Risk Level: Low (default setting preserves previous behaviour)
Testing: Added unit test
Docs Changes: updated API annotation
Release Notes: yes
Platform Specific Features: No
Fixes #10076

### CODE REVIEW

File: api/envoy/extensions/filters/http/lua/v3/lua.proto [hunk 1]
```diff
@@ -78,6 +80,13 @@ message Lua {
   //         stat_prefix: bar_script # This emits lua.bar_script.errors etc.
   //
   string stat_prefix = 4;
+
+  // Determines if HTTP body is buffered when Lua script executes httpCall.
+  // If the setting is false, Envoy will put pressure on the downstream client to stop sending
+  // data. If the setting is true, Envoy will try to buffer entire body sent by downstream client,
+  // but buffering may exceed maximum buffer limit per connection and the connection may be dropped.
+  // Defaults to true.
+  google.protobuf.BoolValue buffer_body = 5;
```
Comment (wbpcode) [inline api/envoy/extensions/filters/http/lua/v3/lua.proto:89]: I am fine to this feature. But the naming `buffer_body` will indicates that this field is used to determine whether the lua filter will buffer the body data.

So, I think `flow_backpressure` or `backpressure` maybe better.

**Comment by htuch** [comment]
@tyxia is in the process of adding similar support for ext_proc and gRPC side-calls via HTTP async client. @tyxia any thoughts on whether that is relevant here?

File: source/extensions/filters/http/lua/lua_filter.cc [hunk 2]
```diff
@@ -278,7 +278,10 @@ Http::FilterDataStatus StreamHandleWrapper::onData(Buffer::Instance& data, bool
     resumeCoroutine(0, yield_callback_);
   }
 
-  if (state_ == State::HttpCall || state_ == State::WaitForBody) {
+  if (state_ == State::HttpCall) {
+    return filter_.config()->bufferBody() ? Http::FilterDataStatus::StopIterationAndBuffer
```
Comment (tyxia) [inline source/extensions/filters/http/lua/lua_filter.cc:282]: Is this correct? Both `StopIterationAndBuffer` and `StopIterationAndWatermark` will buffer the body. 

The difference here is that former one will return 413 like hard limit and the later on will invoke watermark callbacks `onDecoderFilterAboveWriteBufferHighWatermark` for example.

**Comment by tyxia** [comment]
> @tyxia is in the process of adding similar support for ext_proc and gRPC side-calls via HTTP async client. @tyxia any thoughts on whether that is relevant here?

@htuch  This is different from my sidestream flow control.  Let's take a look at this data flow like below for example
```
downstream_client ---> filter(either ext_proc or lua) ----> sideStream server
```

In this PR: the back pressure is generated at filter level when data is buffered using filter manager's watermark buffer.  This is commonly used in body buffered case. We actually have similar back-pressure in ext_proc sync mode already when data is buffered internally.   The `filter---> sidestream server` path is not directly involved here, but it leads to back-pressure in a indirect way when filter buffered data can not be consumed in time, e.g.,  due to slow sidestream server.

In my sidestream flow control: the backpressure is generated at connection/codec watermark buffer on the path `filter---> sidestream server`. (1) It works regardless of whether filter buffers data or not. For example non-buffering case in ext_proc async mode . (2) Also this is direct back-pressure when sidestream server is slow and filter serves as transmission of backpressure from sidestream server to downstream client.

File: source/extensions/filters/http/lua/lua_filter.cc (hunk 2 above)
Comment (cpakulski) [reply source/extensions/filters/http/lua/lua_filter.cc:282]: I think it is correct, unless I am missing something. My expectation is that watermark callbacks will indicate to the listener to stop reading from input, therefore putting backpressure on the sender. The use case here is to read headers, issue httpCall (which may be take some time to complete) and pass the body to upstream receiver without investigating it. The problem is that while Lua filter waits for httpCall to complete, client sends so much data that it exceeds configured maximum and connection is terminated.

Comment (tyxia) [reply source/extensions/filters/http/lua/lua_filter.cc:282]: My point is your check here is if  body is buffered, while `StopIterationAndBuffer ` and `StopIterationAndWatermark ` will buffer the body by FM. I think my point is kind of similar to @wbpcode 's [API review](https://github.com/envoyproxy/envoy/pull/35104#discussion_r1669561250)

The difference between those two is not about buffer body instead it is about how overflow data is handled. StopIterationAndBuffer will result in 413 local reply and StopIterationAndWatermark will result in backpressure by watermark callbacks.

Comment (cpakulski) [reply source/extensions/filters/http/lua/lua_filter.cc:282]: Got it. Thanks. Yeah, the naming may be misleading. When I was going through the code I realized that Lua filter always buffers body, even when the script does not need the body at all, so I was thinking that having a config option to skip waiting for entire body may be useful for httpCall situation and when script does not need the body (possibly future PR). Hence the name `buffer_body`. But I do not have a full context, so decided to limit the change only for httpCall.

File: api/envoy/extensions/filters/http/lua/v3/lua.proto (hunk 1 above)
Comment (cpakulski) [reply api/envoy/extensions/filters/http/lua/v3/lua.proto]: Renamed to `flow_control`. This was great suggestion. Thanks!

File: api/envoy/extensions/filters/http/lua/v3/lua.proto [hunk 3]
```diff
@@ -78,6 +81,13 @@ message Lua {
   //         stat_prefix: bar_script # This emits lua.bar_script.errors etc.
   //
   string stat_prefix = 4;
+
+  // Determines if Envoy will exercise flow control when receiving HTTP body when Lua script executes httpCall.
+  // If the setting is true, Envoy will put backpressure on the downstream client to stop sending
+  // data. If the setting is false, Envoy will not put any backpressure on downstream client
+  // and will buffer HTTP body. Note that buffering may exceed maximum buffer limit per connection and the connection may be dropped.
+  // Defaults to false.
```
Comment (htuch) [inline api/envoy/extensions/filters/http/lua/v3/lua.proto:89]: I guess the main question I'd have API wise is whether this should just be the default behavior (with runtime control)?

File: source/extensions/filters/http/lua/lua_filter.cc (hunk 2 above)
Comment (wbpcode) [reply source/extensions/filters/http/lua/lua_filter.cc]: >  My expectation is that watermark callbacks will indicate to the listener to stop reading from input, therefore putting backpressure on the sender.

Is this ensured that Envoy will do that? If yes then please comment it in the code to let the future developer/reader know that.

File: api/envoy/extensions/filters/http/lua/v3/lua.proto (hunk 3 above)
Comment (cpakulski) [reply api/envoy/extensions/filters/http/lua/v3/lua.proto:89]: That new API flag is purely for backwards compatibility (if it is discovered that it somehow breaks executing Lua scripts). Assuming that is does not break anything, I cannot imagine why a user would choose one option vs the other. It seems that runtime control is a better option. @htuch do you suggest temporary (6 months long) runtime_guard?

Comment (htuch) [reply api/envoy/extensions/filters/http/lua/v3/lua.proto:89]: I'm good with a runtime guard providing folks who know more about the typical Lua use case than I don't think this is going to cause widespread breakage. Ideally we just have the right behavior by default.

**Review by tyxia — DISMISSED** [review]
LGTM

Defer to lua extension owner @wbpcode

**Comment by KBaichoo** [comment]
PTAL @wbpcode

**Review by wbpcode — DISMISSED** [review]

**Comment by wbpcode** [comment]
@cpakulski Hi, could you merge main and resolve the conflict? Then we can merge this PR.

**Comment by cpakulski** [comment]
@wbpcode Thanks for reviewing. Rebased and resolved conflicts.

**Review by wbpcode — APPROVED** [review]

**Pull request merged**